### PR TITLE
power actor invariant tests

### PR DESCRIPTION
**Pull request by acruikshank**
work towards #1165 and #1166 

### Motivation

We would like to be able to check the internal consistency of the storage power actor's state. This PR adds `CheckStateInvariants` for power state and calls it for all unit tests that are not expected to fail. 

### Proposed Changes

1. Add `CheckStateInvariants`
2. Add test harness `checkState` and call it in all relevant unit tests.
3. Add a call to `CheckStateInvariants` to `states.CheckStateInvariants`.
4. Add `CheckMinersAgainstPower` to check cross actor state
5. Run `CheckStateInvariants` in scenario tests including correct_cc_upgrade migration test.
6. Fix issue with power migration where a miner's corrected claim would not update power state in a consistent way.
7. Change account invariant check to only require a public key if actor is non-singleton. (when I rebased in the inclusion of account to the main state check, it wasn't working in scenario tests without this fix).

### CODE REVIEW

File: actors/builtin/power/testing.go
Comment (anorth) [inline actors/builtin/power/testing.go:127]: Ok, I do kinda like this. We could propagate this pattern more widely.

File: actors/builtin/power/testing.go
Comment (anorth) [inline actors/builtin/power/testing.go:154]: You can just use `len(byAddress)`

File: actors/builtin/power/testing.go
Comment (anorth) [inline actors/builtin/power/testing.go:63]: This duplicates line 59

File: actors/builtin/power/testing.go
Comment (anorth) [inline actors/builtin/power/testing.go:79]: Blocking: `msgs` and `err` are both ignored here.

It looks like I've been inconsistent to date, but I think that inner methods within each actor's CheckStateInvariants should receive the accumulator as a parameter, rather than return a new one, to avoid this easy-to-make mistake.

**Review by anorth — COMMENTED** [review]
Looking good

File: actors/migration/power.go
Comment (anorth) [inline actors/migration/power.go:126]: See my comment on #1237, but in any case don't include this fix in this otherwise validation-only PR.

File: actors/builtin/power/testing.go
Comment (anorth) [inline actors/builtin/power/testing.go:58]: Remove this before merging.

File: actors/builtin/power/testing.go
Comment (anorth) [inline actors/builtin/power/testing.go:194]: It is in fact an invariant that this queue is nil at the end of every epoch, because it is emptied by cron.

File: actors/migration/test/correct_cc_upgrade_then_fault_scenario_test.go
Comment (anorth) [inline actors/migration/test/correct_cc_upgrade_then_fault_scenario_test.go:405]: Upon reverting the migration fix above, I guess this won't hold. It's important that the state checks are useful to us as human network operators, at least as much as they are useful for tests. After the upgrade, we need to be able to run the state checks and see the un-fixed mismatch. That means this test must tolerate non-empty errors.

I think change this to assert the expected error message/s are returned.

File: actors/states/check.go
Comment (anorth) [inline actors/states/check.go:206]: Can you print _both_ epochs?

**Review by anorth — APPROVED** [review]
LGTM after reverting the fix that's in #1237.

File: actors/builtin/power/testing.go
Comment (acruikshank) [reply actors/builtin/power/testing.go:194]: Agreed, but the tests will sometimes run this before cron, so it might catch something.

File: actors/migration/power.go
Comment (acruikshank) [reply actors/migration/power.go]: reverted

**Pull request merged**
